AI/VCAI/HeroManager.cpp

```diff
@@ -0,0 +1,237 @@
+		std::make_shared<AtLeastOneMagicRule>()
+	});
+
+SecondarySkillEvaluator HeroManager::scountSkillsScores = SecondarySkillEvaluator(
+	{
+		std::make_shared<SecondarySkillScoreMap>(
+			std::map<SecondarySkill, float>
+			{
+				{SecondarySkill::LOGISTICS, 2},
+				{SecondarySkill::ESTATES, 2},
+				{SecondarySkill::PATHFINDING, 1},
+				{SecondarySkill::SCHOLAR, 1}
+			}),
+		std::make_shared<ExistingSkillRule>()
+	});
+
+void HeroManager::init(CPlayerSpecificInfoCallback * CB)
+{
+	cb = CB;
+}
+
+void HeroManager::setAI(VCAI * AI)
+{
+	ai = AI;
+}
+
+float HeroManager::evaluateSpeciality(const CGHeroInstance * hero) const
+{
+	auto heroSpecial = Selector::source(Bonus::HERO_SPECIAL, hero->type->ID.getNum());
+	auto secondarySkillBonus = Selector::type(Bonus::SECONDARY_SKILL_PREMY);
+	auto specialSecondarySkillBonuses = hero->getBonuses(heroSpecial.And(secondarySkillBonus));
+	float specialityScore = 0.0f;
+
+	for(auto bonus : *specialSecondarySkillBonuses)
+	{
+		SecondarySkill bonusSkill = SecondarySkill(bonus->subtype);
+		float bonusScore = wariorSkillsScores.evaluateSecSkill(hero, bonusSkill);
+
+		if(bonusScore > 0)
+			specialityScore += bonusScore * bonusScore * bonusScore;
+	}
+
+	return specialityScore;
+}
+
+float HeroManager::evaluateFightingStrength(const CGHeroInstance * hero) const
+{
+	return evaluateSpeciality(hero) + wariorSkillsScores.evaluateSecSkills(hero) + hero->level * 1.5f;
+}
+
+std::map<HeroPtr, HeroRole> HeroManager::getHeroRoles() const
+{
+	std::map<HeroPtr, float> scores;
+	std::map<HeroPtr, HeroRole> result;
+	auto myHeroes = ai->getMyHeroes();
+
+	for(auto & hero : myHeroes)
+	{
+		scores[hero] = evaluateFightingStrength(hero.get());
+	}
+
+	std::sort(myHeroes.begin(), myHeroes.end(), [&](const HeroPtr & h1, const HeroPtr & h2) -> bool
+	{
+		return scores.at(h1) < scores.at(h2);
+	});
+
+	int mainHeroCount = 4;
+
+	for(auto & hero : myHeroes)
+	{
+		result[hero] = (mainHeroCount--) > 0 ? HeroRole::MAIN : HeroRole::SCOUT;
+	}
+
+	return result;
+}
+
+int HeroManager::selectBestSkill(const HeroPtr & hero, const std::vector<SecondarySkill> & skills) const
+{
+	auto roles = getHeroRoles();
+	auto role = roles[hero];
+	auto & evaluator = role == HeroRole::MAIN ? wariorSkillsScores : scountSkillsScores;
+
+	int result = 0;
+	float resultScore = -100;
+
+	for(int i = 0; i < skills.size(); i++)
+	{
+		auto score = evaluator.evaluateSecSkill(hero.get(), skills[i]);
+
+		if(score > resultScore)
+		{
+			resultScore = score;
+			result = i;
+		}
+
+		logAi->trace(
+			"Hero %s is proposed to learn %d with score %f",
+			hero.name,
+			skills[i].toEnum(),
+			score);
+	}
+
+	return result;
+}
+
+SecondarySkillScoreMap::SecondarySkillScoreMap(std::map<SecondarySkill, float> scoreMap)
+	:scoreMap(scoreMap)
+{
+}
+
+void SecondarySkillScoreMap::evaluateScore(const CGHeroInstance * hero, SecondarySkill skill, float & score) const
+{
+	auto it = scoreMap.find(skill);
+
+	if(it != scoreMap.end())
+	{
+		score = it->second;
```

Comment:
You can use our ```vstd``` function in situations like this:
```
	if(vstd::contains(scoreMap, skill))
	{
		score = scoreMap[skill];
```

Reply:
Using iterator in this case is faster, `vstd::contains` should be used if you just check presence and nothing more.

Reply:
I think I will fix it soon

Reply:
I thought it has to be fixed but it turned out that it was already correct. To avoid double lookup in map I used iterator. So this can be merged.